### ISSUE TITLE
Feature request: random offset field

### ISSUE THREAD

**Issue by paulevsGitch** [minetest/minetest#8262]
##### Issue type
 - Feature request
 
My suggestion is to add random offset field for nodes (as vector with separate x, y and z values, which are multipliers for it), used for offset node on render (random must be based on node position in the world). This can be used for many decorative nods, like plants, butterflies, etc.

Example vizualisation:
- No offset
![grass_a](https://user-images.githubusercontent.com/17707380/53024927-13399380-3471-11e9-87b8-82d3dae4c667.jpg)
- Offset in XZ plane
![grass_b](https://user-images.githubusercontent.com/17707380/53024928-13399380-3471-11e9-863d-eb5d7fd2a1aa.jpg)

**Comment by NathanSalapat** [comment]
The new plantlike meshoptions already provide an option to offset placement, though it is random. There was discussion about this before as well though I can't find the issue.

**Comment by paulevsGitch** [comment]
> The new plantlike meshoptions already provide an option to offset placement, though it is random. There was discussion about this before as well though I can't find the issue.
> Relevant text in API.
> https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L978

Thank you for answer, but as I see - it is only horizontal and can be used only with plantlike draw type. I think, if it will be controllable and can be used with other drawing types (like mesh or node box for example) it will be better.

**Comment by paramat** [comment]
Meshoptions has vertical offsets too.

There may be other drawtypes that benefit from having random offsets, but it seems low priority as plantlike is the obvious one. Nodeboxes are very rarely used in ways that would benefit from random offsets (for example used as plants).
It can't be done for basic cubic nodes as the lightweight rendering and lighting for those depends on being grid-aligned.

So what drawtypes do you suggest? An obvious one is plantlike_rooted.

Keep in mind this is a voxel game engine, which means an abstract world where most objects are rigidly aligned to a grid. Random offsets are therefore somewhat out of place in a voxel engine. They are also more intensive to process and remove some benefits that come with alignment.
If alignment bothers someone a lot they are simply using the wrong game =)

**Comment by paulevsGitch** [comment]
> Meshoptions has vertical offsets too.
> 
> There may be other drawtypes that benefit from having random offsets, but it seems low priority as plantlike is the obvious one. Nodeboxes are very rarely used in ways that would benefit from random offsets (for example used as plants).
> It can't be done for basic cubic nodes as the lightweight rendering and lighting for those depends on being grid-aligned.
> 
> So what drawtypes do you suggest? An obvious one is plantlike_rooted.
> 
> Keep in mind this is a voxel game engine, which means an abstract world where most objects are rigidly aligned to a grid. Random offsets are therefore somewhat out of place in a voxel engine. They are also more intensive to process and remove some benefits that come with alignment.
> If alignment bothers someone a lot they are simply using the wrong game =)
> 
> If a node is offset, i guess this is clientside only? If so, the collisionboxes will not be offset, which will be weird for some nodes like nodeboxes.

Offsets for mesh nodes are one of desired function, it can be used for plenty objects (small stones, custom modeled plant, another decorative stuff). Also, I suggest use vector as multiplier, so you can control how large offset on each direction your node will have.

About collision boxes - I think, that they may be the same, as without offset (so offset will be client side only)

**Comment by Wuzzy2** [comment]
I have to agree with paramat. It kind of defeats the whole promise of Minetest.

Especially things will get weird and more annoying with regards to selectionboxes. I love pixel-perfect selectionboxes. I hate selectionboxes that are way off.

**Comment by paulevsGitch** [comment]
Not everybody likes only grid-attached objects. For world decorations, that tries to mimic nature, it looks boring, like grids of flowers or butterflies. I agree, that solid nodes must be aligned to grid, but why meshlike or other nodes must be attached to it also? Controllable offset is a great possibility to decor.

**Comment by paramat** [comment]
Yes due to collisionboxes not being offset, these clientside visual offsets are only useful for things that usually do not cause collisions, so mostly small plants.
'plantlike_rooted' drawtype is valid for random offsets, but there is already an issue that requests random offsets for that #7937 

It's a case of whether the usefulness justifies the work, extra code and maintenance. Non-colliding decorations (usually small plants) overwhelmingly use plantlike, not meshnodes or nodeboxes which are far more intensive to render in large numbers, so are problematic for use as high density decorations.

Note that unlike plantlike, both meshnodes and nodeboxes can be defined multiple times, each with a different built-in offset, and the decoration API can randomly place them, so there is a workaround. They would also collide properly.

Decorations don't try to mimic nature, they are a grid-aligned abstraction. If someone doesn't like grid-aligned objects or finds them boring they are using the wrong game engine. The grid is the fundamental aspect of the whole game engine.
